Zend/zend_fibers.c

```diff
@@ -0,0 +1,719 @@
+		if (previous) {
+			zend_fiber_suspend(previous); // Still in fiber, suspend again until in {main}.
+			abort(); // This fiber should never be resumed.
+		}
+
+		zend_fiber_error *error = EG(fiber_error);
+		zend_error_zstr_at(error->type, error->filename, error->lineno, error->message);
+	}
+}
+
+ZEND_COLD void zend_error_suspend_fiber(
+		int orig_type, const char *error_filename, uint32_t error_lineno, zend_string *message)
+{
+	ZEND_ASSERT(EG(current_fiber) && "Must be within an active fiber!");
+
+	zend_fiber_error *error = emalloc(sizeof(zend_fiber_error));
+
+	error->type = orig_type;
+	error->filename = error_filename;
+	error->lineno = error_lineno;
+	error->message = message;
+
+	EG(fiber_error) = error;
+
+	zend_fiber_suspend(EG(current_fiber));
+
+	abort(); // This fiber should never be resumed.
+}
+
+static zend_always_inline zend_vm_stack zend_fiber_vm_stack_alloc(size_t size)
+{
+	zend_vm_stack page = emalloc(size);
+
+	page->top = ZEND_VM_STACK_ELEMENTS(page);
+	page->end = (zval *) ((uintptr_t) page + size);
+	page->prev = NULL;
+
+	return page;
+}
+
+static void ZEND_STACK_ALIGNED zend_fiber_execute(zend_fiber_context *context)
+{
+	zend_fiber *fiber = EG(current_fiber);
+	ZEND_ASSERT(fiber);
+
+	zend_long error_reporting = INI_INT("error_reporting");
+	if (!error_reporting && !INI_STR("error_reporting")) {
+		error_reporting = E_ALL;
+	}
+
+	zend_vm_stack stack = zend_fiber_vm_stack_alloc(ZEND_FIBER_VM_STACK_SIZE);
+	EG(vm_stack) = stack;
+	EG(vm_stack_top) = stack->top + ZEND_CALL_FRAME_SLOT;
+	EG(vm_stack_end) = stack->end;
+	EG(vm_stack_page_size) = ZEND_FIBER_VM_STACK_SIZE;
+
+	fiber->execute_data = (zend_execute_data *) stack->top;
+
+	memset(fiber->execute_data, 0, sizeof(zend_execute_data));
+
+	EG(current_execute_data) = fiber->execute_data;
+	EG(jit_trace_num) = 0;
+	EG(error_reporting) = error_reporting;
+
+	fiber->fci.retval = &fiber->value;
+
+	// Reference added while running, removed when suspended, and added again once resumed.
+	GC_ADDREF(&fiber->std);
+
+	fiber->status = ZEND_FIBER_STATUS_RUNNING;
+
+	zend_call_function(&fiber->fci, &fiber->fci_cache);
+
+	if (EG(exception)) {
+		if (fiber->status == ZEND_FIBER_STATUS_SHUTDOWN) {
+			if (EXPECTED(zend_is_graceful_exit(EG(exception)) || zend_is_unwind_exit(EG(exception)))) {
+				zend_clear_exception();
+			} else {
+				zend_exception_error(EG(exception), E_ERROR);
+			}
+		} else {
+			fiber->status = ZEND_FIBER_STATUS_THREW;
+		}
+	} else {
+		fiber->status = ZEND_FIBER_STATUS_RETURNED;
+	}
+
+	zend_vm_stack_destroy();
+	fiber->execute_data = NULL;
+
+	// Remove reference added at last resume.
+	GC_DELREF(&fiber->std);
+}
+
+static zend_object *zend_fiber_object_create(zend_class_entry *ce)
+{
+	zend_fiber *fiber;
+
+	fiber = emalloc(sizeof(zend_fiber));
+	memset(fiber, 0, sizeof(zend_fiber));
+
+	zend_object_std_init(&fiber->std, ce);
+	fiber->std.handlers = &zend_fiber_handlers;
+
+	zend_hash_index_add_ptr(&EG(fibers), (uintptr_t) fiber, fiber);
```

Comment:
I believe you can now drop the `EG(fibers)` table. It was only used for shutdown.

Reply:
Yeah, only reason would be if we wanted to expose a list of running fibers for debugging, but we can always add that in the future. I'll remove it.